akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedBehavior.scala

```diff
@@ -0,0 +1,113 @@
+
+import akka.actor.NoSerializationVerificationNeeded
+import akka.actor.typed.Behavior
+import akka.actor.typed.Behavior.StoppedBehavior
+import akka.actor.typed.scaladsl.{ ActorContext, TimerScheduler }
+import akka.annotation.InternalApi
+import akka.event.{ LogSource, Logging }
+import akka.persistence.typed.scaladsl.PersistentBehaviors
+import akka.persistence.{ JournalProtocol, Persistence, RecoveryPermitter, SnapshotProtocol }
+import akka.{ actor ⇒ a }
+
+/** INTERNAL API */
+@InternalApi
+private[akka] object EventsourcedBehavior {
+
+  // we share the same counter with untyped, in order to not accidentally overlap, since untyped can co-exist with typed
+  private[akka] val instanceIdCounter = akka.persistence.Eventsourced.instanceIdCounter
+
+  @InternalApi private[akka] object WriterIdentity {
+    def newIdentity(): WriterIdentity = {
+      val instanceId: Int = EventsourcedBehavior.instanceIdCounter.getAndIncrement()
+      val writerUuid: String = UUID.randomUUID.toString
+      WriterIdentity(instanceId, writerUuid)
+    }
+  }
+  @InternalApi private[akka] final case class WriterIdentity(instanceId: Int, writerUuid: String)
+
+  /** INTERNAL API: Protocol used internally by the eventsourced behaviors, ever exposed to user-land */
+  @InternalApi private[akka] sealed trait EventsourcedProtocol
+  @InternalApi private[akka] case object RecoveryPermitGranted extends EventsourcedProtocol
+  @InternalApi private[akka] final case class JournalResponse(msg: akka.persistence.JournalProtocol.Response) extends EventsourcedProtocol
+  @InternalApi private[akka] final case class SnapshotterResponse(msg: akka.persistence.SnapshotProtocol.Response) extends EventsourcedProtocol
+  @InternalApi private[akka] final case class RecoveryTickEvent(snapshot: Boolean) extends EventsourcedProtocol
+  @InternalApi private[akka] final case class ReceiveTimeout(timeout: akka.actor.ReceiveTimeout) extends EventsourcedProtocol
+
+  /** INTERNAL DSL */
+  private[akka] implicit class UnlessStopped[T](val b: Behavior[T]) extends AnyVal {
+    /** Returns `other` behavior if the wrapped one is `Stopped`. */
+    def unlessStopped(other: ⇒ Behavior[T]): Behavior[T] =
+      b match {
+        case _: StoppedBehavior[T @unchecked] ⇒ b // still the stopped one
+        case _                                ⇒ other
+      }
+  }
+
+  implicit object PersistentBehaviorLogSource extends LogSource[EventsourcedBehavior[_, _, _]] {
+    override def genString(b: EventsourcedBehavior[_, _, _]): String = {
+      val behaviorShortName = b match {
+        case _: EventsourcedRunning[_, _, _]                  ⇒ "running"
+        case _: EventsourcedRecoveringEvents[_, _, _]         ⇒ "recover-evts"
+        case _: EventsourcedRecoveringSnapshot[_, _, _]       ⇒ "recover-snap"
+        case _: EventsourcedRequestingRecoveryPermit[_, _, _] ⇒ "awaiting-permit"
+      }
+      s"PersistentBehavior[id:${b.persistenceId}][${b.context.self.path}][$behaviorShortName]"
```

Comment:
WDYT about the tostring? I would have liked to keep the behavior short name... but maybe coding it as `S|RE|RS|R` wound be better? it gets a bit long. Definitely want to keep the persistence id and actor name in it though, otherwise it's a mess to try to find out what logged what

Reply:
Example:

```

[INFO] [02/02/2018 18:24:35.991] [PersistentBehaviorSpec-akka.actor.default-dispatcher-9] [PersistentBehavior[id:c12][akka://PersistentBehaviorSpec/user/$w][awaiting-permit]] [c12][$w] Becoming recovering SNAPSHOT: Recovery(SnapshotSelectionCriteria(9223372036854775807,9223372036854775807,0,0),9223372036854775807,9223372036854775807)
[INFO] [02/02/2018 18:24:35.992] [PersistentBehaviorSpec-akka.actor.default-dispatcher-9] [PersistentBehavior[id:c12][akka://PersistentBehaviorSpec/user/$w][recover-snap]] LOAD SNAPSHOT: SnapshotSelectionCriteria(9223372036854775807,9223372036854775807,0,0)
```